src/ir/opt/binarize.rs

```diff
@@ -0,0 +1,120 @@
+//! Binarize terms
+
+use crate::ir::term::*;
+
+/// Binarize cache.
+#[derive(Default)]
+pub struct Cache(TermMap<Term>);
+
+impl Cache {
+    /// Empty cache.
+    pub fn new() -> Self {
+        Cache(TermMap::new())
+    }
+}
+
+fn binarize(op: &Op, children: &[Term]) -> Term {
+    if children.is_empty() || children.len() == 1 {
+        term(op.clone(), children.to_vec())
+    } else if children.len() == 2 {
+        return term![op.clone(); children[0].clone(), children[1].clone()];
+    } else {
+        return term![op.clone(); children[0].clone(), binarize(op, &children[1..].to_vec())];
+    }
+}
```

Comment:
I'd recommend using `Iterator::fold` instead of recursion here. We build some *very* large n-ary terms, and don't want to blow the stack.